x-pack/plugins/triggers_actions_ui/public/application/components/json_editor_with_message_variables.tsx

```diff
@@ -44,29 +63,83 @@ export const JsonEditorWithMessageVariables: React.FunctionComponent<Props> = ({
   helpText,
   onBlur,
 }) => {
-  const [cursorPosition, setCursorPosition] = useState<any>(null);
+  const editorRef = useRef<monaco.editor.IStandaloneCodeEditor>();
+  const editorDisposables = useRef<monaco.IDisposable[]>([]);
+  const [showErrorMessage, setShowErrorMessage] = useState(false);
 
   const { convertToJson, setXJson, xJson } = useXJsonMode(inputTargetValue ?? null);
 
   const onSelectMessageVariable = (variable: ActionVariable) => {
+    const editor = editorRef.current;
+    if (!editor) {
+      setShowErrorMessage(true);
+      return;
+    }
+    const cursorPosition = editor.getSelection();
     const templatedVar = templateActionVariable(variable);
+
     let newValue = '';
     if (cursorPosition) {
-      const cursor = cursorPosition.getCursor();
-      cursorPosition.session.insert(cursor, templatedVar);
-      newValue = cursorPosition.session.getValue();
+      editor.executeEdits(EDITOR_SOURCE, [
+        {
+          range: cursorPosition,
+          text: templatedVar,
+        },
+      ]);
+      newValue = editor.getValue();
     } else {
       newValue = templatedVar;
     }
+    setShowErrorMessage(false);
     setXJson(newValue);
     // Keep the documents in sync with the editor content
     onDocumentsChange(convertToJson(newValue));
   };
 
-  const onClickWithMessageVariable = (_value: any) => {
-    setCursorPosition(_value);
+  const registerEditorListeners = useCallback(() => {
+    const editor = editorRef.current;
+    if (!editor) {
+      return;
+    }
+    editorDisposables.current.push(
+      editor.onDidBlurEditorWidget(() => {
+        onBlur?.();
+      })
+    );
+  }, [onBlur]);
+
+  const unregisterEditorListeners = () => {
```

Comment:
Is there example code somewhere for how to use this new editor? I tried following the issue trail but didn't see much in terms of "this is how you implement it"

Reply:
there is some precedence for this pattern in the lens plugin `formula_editor.tsx`

Reply:
Gotcha, I don't really know how to review this part honestly. It seems okay but is it worth reaching out to folks who know more about the EUI CodeEditor to make sure this is looks right (no edge cases unresolved, no memory leaks, etc)?

Reply:
@jloleysens Hey there, not sure if you're the right person I should be tagging, but just wanted to get you in this PR to get your opinion with respects to using this editor properly, especially regarding disposing of the monaco editor event listeners. Thanks!